capirca/lib/fortigate.py

```diff
@@ -0,0 +1,330 @@
+    389: 'LDAP',
+    69: 'TFTP'
+
+  }
+
+  _PORTS_UDP = {
+    53: 'DNS',
+    7: 'PING',
+    500: 'IKE',
+    2049: 'NFS',
+    123: 'NTP',
+    520: 'RIP',
+    161: 'SNMP',
+    162: 'snmptrap',
+    514: 'SYSLOG',
+    517: 'TALK',
+    69: 'TFTP',
+    37: 'TIMESTAMP',
+    1812: 'RADIUS',
+    67: 'DHCP'
+
+  }
+
+  _PROTO_MAP = {
+    'icmp': 'ALL_ICMP',
+    'gre': 'GRE',
+    'ip': 'ALL',
+    'tcp': _PORTS_TCP,
+    'udp': _PORTS_UDP
+  }
+
+  @staticmethod
+  def GetProtocol(protocol, port=None):
+    f_proto = FortigatePortMap._PROTO_MAP.get(protocol, None)
+    if f_proto is None:
+      raise FortiGateValueError('%r protocol is not supported by Fortigate, supported protocols = %r' % (
+        protocol, FortigatePortMap._PROTO_MAP.keys()))
+
+    if isinstance(f_proto, str):
+      return f_proto
+    elif port:
+      return f_proto[port]
+
+    else:
+      raise FortiGateFindServiceError('failed to get service from %r protocol and %r port' % (protocol, port))
+
+
+class Term(aclgenerator.Term):
+  ALLOWED_PROTO_STRINGS = ['gre', 'icmp', 'ip', 'tcp', 'udp']
+  COMMENT_MAX_WIDTH = 70
+
+  FW_ADDRESSES = []
+  FW_SERVICES = []
+
+  _FW_DUP_CHECK = set()
+
+  CURRENT_ID = 0
+
+  def __init__(self, term):
+    super(Term, self).__init__(term)
+    self._term = term
+
+    self.id = type(self).CURRENT_ID
+    type(self).CURRENT_ID += 1
+
+  @staticmethod
+  def get_fw_addresses():
+    Term.FW_ADDRESSES.extend([' ', 'end', ' '])
+    return Term.FW_ADDRESSES
+
+  @staticmethod
+  def get_fw_services():
+    Term.FW_SERVICES.extend([' ', 'end', ' '])
+    return Term.FW_SERVICES
+
+  @staticmethod
+  def _get_addresses_name(addresses):
+    v4_addresses = [x.with_prefixlen for x in addresses if
+                    not isinstance(x, nacaddr.IPv6)]
+    addresses = ' '.join(v4_addresses)
+    return addresses or 'all'
+
+  def _get_services_string(self, protocol, ports):
+
+    services = []
+    if protocol and not ports:
+      services.append(FortigatePortMap.GetProtocol(protocol[0]))
+    for port in ports:
+      try:
+        service = FortigatePortMap.GetProtocol(protocol[0], port[0])
+      except KeyError:
+        self._add_service_to_fw_services(protocol[0], port[0])
+        service = str(port[0])
+      services.append(service)
+
+    return ' '.join(services) or 'ALL'
+
+  def _add_address_to_fw_addresses(self, addr):
+    if addr in type(self)._FW_DUP_CHECK:
+      return
+    type(self).FW_ADDRESSES.extend(['\tedit %s' % addr,
+                                    '\t\tset subnet %s' % addr,
+                                    '\tnext'])
+    type(self)._FW_DUP_CHECK.add(addr)
+
+  def _add_service_to_fw_services(self, protocol, service):
+    if service in type(self)._FW_DUP_CHECK:
+      return
+
+    type(self).FW_SERVICES.extend(['\tedit %s' % service,
+                                   '\t\tset protocol TCP/UDP',
+                                   '\t\tset %s-portrange %s' % (protocol.lower(), service),
```

Comment:
Line too long, you can fix this by moving the list to its own new line and indenting 
```
    type(self).FW_SERVICES.extend(
        ['\tedit %s' % service,
         '\t\tset protocol TCP/UDP',
         '\t\tset %s-portrange %s' % (protocol.lower(), service),
         '\tnext']
        )
```